File: BensBible2/ViewModels/MemorizeViewModel.swift
import Foundation
import SwiftData

@Observable
final class MemorizeViewModel {
    private var modelContext: ModelContext?

    var memorizedVerses: [MemorizedVerse] = []
    var dueVerses: [MemorizedVerse] = []

    // Session state
    var sessionQueue: [MemorizedVerse] = []
    var sessionIndex: Int = 0
    var sessionComplete: Bool = false
    var sessionCorrect: Int = 0
    var sessionTotal: Int = 0
    var currentExercise: ExerciseState?
    var showingResult: Bool = false
    var lastResultWasCorrect: Bool = false
    private var lastExerciseType: ExerciseType?

    func configure(modelContext: ModelContext) {
        self.modelContext = modelContext
        load()
    }

    func load() {
        guard let context = modelContext else { return }
        let descriptor = FetchDescriptor<MemorizedVerse>(sortBy: [SortDescriptor(\.addedAt)])
        memorizedVerses = (try? context.fetch(descriptor)) ?? []
        dueVerses = memorizedVerses.filter { $0.isDue }
    }

    func addVerse(bookName: String, chapterNumber: Int, verseNumber: Int, verseText: String) {
        guard let context = modelContext else { return }
        let key = "\(bookName) \(chapterNumber):\(verseNumber)"
        guard !memorizedVerses.contains(where: { $0.verseKey == key }) else { return }
        let verse = MemorizedVerse(bookName: bookName, chapterNumber: chapterNumber, verseNumber: verseNumber, verseText: verseText)
        context.insert(verse)
        load()
    }

    func removeVerse(_ verse: MemorizedVerse) {
        guard let context = modelContext else { return }
        context.delete(verse)
        load()
    }

    func startSession() {
        sessionQueue = dueVerses.shuffled()
        sessionIndex = 0
        sessionCorrect = 0
        sessionTotal = 0
        sessionComplete = false
        lastExerciseType = nil
        advanceToNextExercise()
    }

    func submitAnswer(quality: Int) {
        guard let context = modelContext,
              let exercise = currentExercise else { return }
        let verse = exercise.verse

        lastResultWasCorrect = quality >= 3
        sessionTotal += 1
        if quality >= 4 { sessionCorrect += 1 }

        let log = MemoryReviewLog(verseKey: verse.verseKey, quality: quality, exerciseType: exercise.exerciseType.rawValue)
        context.insert(log)

        let result = SM2Scheduler.process(
            quality: quality,
            repetitions: verse.repetitions,
            easeFactor: verse.easeFactor,
            intervalDays: verse.intervalDays
        )
        verse.repetitions = result.repetitions
        verse.easeFactor = result.easeFactor
        verse.intervalDays = result.intervalDays
        verse.nextReviewDate = result.nextReviewDate
        verse.lastReviewedAt = Date()
        verse.totalReviews += 1

        lastExerciseType = exercise.exerciseType
        showingResult = true

        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
            self?.showingResult = false
            self?.moveToNext()
        }
    }

    func skipVerse() {
        moveToNext()
    }

    // MARK: - Private

    private func moveToNext() {
        sessionIndex += 1
        if sessionIndex >= sessionQueue.count {
            sessionComplete = true
            currentExercise = nil
            load()
        } else {
            advanceToNextExercise()
        }
    }

    private func advanceToNextExercise() {
        guard sessionIndex < sessionQueue.count else {
            sessionComplete = true
            return
        }
        let verse = sessionQueue[sessionIndex]
        currentExercise = generateExercise(for: verse)
    }

    private func generateExercise(for verse: MemorizedVerse) -> ExerciseState {
        let type = selectExerciseType(for: verse)
        lastExerciseType = type
        switch type {
        case .multipleChoice: return makeMultipleChoice(verse: verse)
        case .fillBlank: return makeFillBlank(verse: verse)
        case .wordDrag: return makeWordDrag(verse: verse)
        }
    }

    private func selectExerciseType(for verse: MemorizedVerse) -> ExerciseType {
        if verse.repetitions == 0 { return .multipleChoice }
        if verse.repetitions == 1 { return .fillBlank }
        let all: [ExerciseType] = [.fillBlank, .wordDrag, .multipleChoice]
        let available = lastExerciseType.map { last in all.filter { $0 != last } } ?? all
        return available.randomElement() ?? .fillBlank
    }

    private func makeFillBlank(verse: MemorizedVerse) -> ExerciseState {
        let words = verse.verseText.components(separatedBy: " ")
        let count = words.count
        let blankCount = count < 10 ? 1 : count < 20 ? 2 : 3
        let step = max(1, count / (blankCount + 1))

        var blankIndices = Set<Int>()
        for i in 1...blankCount {
            var idx = min(i * step, count - 1)
            for offset in 0..<5 {
                let candidate = (idx + offset) % count
                let stripped = words[candidate].filter { $0.isLetter }
                if stripped.count > 3 { idx = candidate; break }
            }
            blankIndices.insert(idx)
        }

        let segments: [FillBlankSegment] = words.enumerated().map { i, word in
            blankIndices.contains(i) ? .blank(answer: word) : .word(word)
        }
        return .fillBlank(verse: verse, segments: segments)
    }

    private func makeWordDrag(verse: MemorizedVerse) -> ExerciseState {
        let words = verse.verseText.components(separatedBy: " ")
        var shuffled = words.shuffled()
        if shuffled == words && words.count > 1 { shuffled = words.shuffled() }
        return .wordDrag(verse: verse, shuffledWords: shuffled, correctOrder: words)
    }

    private func makeMultipleChoice(verse: MemorizedVerse) -> ExerciseState {
        let distractors = memorizedVerses
            .filter { $0.id != verse.id }
            .shuffled()
            .prefix(3)
            .map { $0.verseText }
        var options = [verse.verseText] + distractors
        let fallbacks = [
            "For the wages of sin is death; but the gift of God is eternal life through Jesus Christ our Lord.",
            "For all have sinned, and come short of the glory of God;",
            "That if thou shalt confess with thy mouth the Lord Jesus, and shalt believe in thine heart that God hath raised him from the dead, thou shalt be saved."
        ]
        var fi = 0
        while options.count < 4 { options.append(fallbacks[fi % fallbacks.count]); fi += 1 }
        return .multipleChoice(verse: verse, options: Array(options.prefix(4)).shuffled())
    }

    func seedDefaultVersesIfNeeded() {
        guard let context = modelContext else { return }
        let descriptor = FetchDescriptor<MemorizedVerse>()
        let count = (try? context.fetch(descriptor).count) ?? 0
        guard count == 0 else { return }

        let defaults: [(String, Int, Int, String)] = [
            ("John", 3, 16, "For God so loved the world, that he gave his only begotten Son, that whosoever believeth in him should not perish, but have everlasting life."),
            ("Jeremiah", 29, 11, "For I know the thoughts that I think toward you, saith the LORD, thoughts of peace, and not of evil, to give you an expected end."),
            ("Philippians", 4, 13, "I can do all things through Christ which strengtheneth me."),
            ("Romans", 8, 28, "And we know that all things work together for good to them that love God, to them who are the called according to his purpose."),
            ("Proverbs", 3, 5, "Trust in the LORD with all thine heart; and lean not unto thine own understanding."),
            ("Isaiah", 40, 31, "But they that wait upon the LORD shall renew their strength; they shall mount up with wings as eagles; they shall run, and not be weary; and they shall walk, and not faint."),
            ("Joshua", 1, 9, "Have not I commanded thee? Be strong and of a good courage; be not afraid, neither be thou dismayed: for the LORD thy God is with thee whithersoever thou goest.")
        ]

        for (book, chapter, verse, text) in defaults {
            let mv = MemorizedVerse(bookName: book, chapterNumber: chapter, verseNumber: verse, verseText: text)
            mv.nextReviewDate = Date()
            context.insert(mv)
        }
        load()
    }
}

File: android/app/src/main/java/com/bensbible/app/viewmodel/MemorizeViewModel.kt
package com.bensbible.app.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bensbible.app.data.MemorizeRepository
import com.bensbible.app.data.MemorizedVerseEntity
import com.bensbible.app.model.ExerciseState
import com.bensbible.app.model.ExerciseType
import com.bensbible.app.model.FillBlankSegment
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class MemorizeViewModel(
    private val repository: MemorizeRepository
) : ViewModel() {

    val allVerses: StateFlow<List<MemorizedVerseEntity>> = repository.getAllVerses()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val dueVerses: StateFlow<List<MemorizedVerseEntity>> = repository.getAllVerses()
        .map { list -> list.filter { it.isDue } }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    // Session state
    var sessionActive by mutableStateOf(false)
        private set
    var currentExercise by mutableStateOf<ExerciseState?>(null)
        private set
    var sessionComplete by mutableStateOf(false)
        private set
    var sessionCorrect by mutableStateOf(0)
        private set
    var sessionTotal by mutableStateOf(0)
        private set
    var showingResult by mutableStateOf(false)
        private set
    var lastResultWasCorrect by mutableStateOf(false)
        private set

    var sessionQueueSize by mutableStateOf(0)
        private set
    var sessionIndex by mutableStateOf(0)
        private set
    private var sessionQueue: List<MemorizedVerseEntity> = emptyList()
    private var lastExerciseType: ExerciseType? = null

    fun addVerse(bookName: String, chapterNumber: Int, verseNumber: Int, verseText: String) {
        viewModelScope.launch {
            repository.addVerse(bookName, chapterNumber, verseNumber, verseText)
        }
    }

    fun deleteVerse(verse: MemorizedVerseEntity) {
        viewModelScope.launch {
            repository.deleteVerse(verse)
        }
    }

    fun startSession(due: List<MemorizedVerseEntity>) {
        sessionQueue = due.shuffled()
        sessionQueueSize = sessionQueue.size
        sessionIndex = 0
        sessionCorrect = 0
        sessionTotal = 0
        sessionComplete = false
        lastExerciseType = null
        sessionActive = true
        advanceToNextExercise()
    }

    fun submitAnswer(quality: Int) {
        val exercise = currentExercise ?: return
        val verse = exercise.verse

        lastResultWasCorrect = quality >= 3
        sessionTotal++
        if (quality >= 4) sessionCorrect++

        val (exerciseTypeName, exerciseTypeEnum) = when (exercise) {
            is ExerciseState.FillBlank -> ExerciseType.FILL_BLANK.name to ExerciseType.FILL_BLANK
            is ExerciseState.WordDrag -> ExerciseType.WORD_DRAG.name to ExerciseType.WORD_DRAG
            is ExerciseState.MultipleChoice -> ExerciseType.MULTIPLE_CHOICE.name to ExerciseType.MULTIPLE_CHOICE
        }
        lastExerciseType = exerciseTypeEnum

        viewModelScope.launch {
            repository.applyReview(verse, quality, exerciseTypeName)
        }

        showingResult = true
        viewModelScope.launch {
            delay(1500)
            showingResult = false
            moveToNext()
        }
    }

    fun skipVerse() {
        moveToNext()
    }

    fun endSession() {
        sessionActive = false
        sessionComplete = false
        currentExercise = null
    }

    // MARK: - Private

    private fun moveToNext() {
        val next = sessionIndex + 1
        sessionIndex = next
        if (next >= sessionQueue.size) {
            sessionComplete = true
            currentExercise = null
        } else {
            advanceToNextExercise()
        }
    }

    private fun advanceToNextExercise() {
        val verse = sessionQueue.getOrNull(sessionIndex) ?: run {
            sessionComplete = true
            return
        }
        currentExercise = generateExercise(verse, allVerses.value)
    }

    private fun generateExercise(
        verse: MemorizedVerseEntity,
        allVerses: List<MemorizedVerseEntity>
    ): ExerciseState {
        val type = selectExerciseType(verse)
        lastExerciseType = type
        return when (type) {
            ExerciseType.MULTIPLE_CHOICE -> makeMultipleChoice(verse, allVerses)
            ExerciseType.FILL_BLANK -> makeFillBlank(verse)
            ExerciseType.WORD_DRAG -> makeWordDrag(verse)
        }
    }

    private fun selectExerciseType(verse: MemorizedVerseEntity): ExerciseType {
        return when (verse.repetitions) {
            0 -> ExerciseType.MULTIPLE_CHOICE
            1 -> ExerciseType.FILL_BLANK
            else -> {
                val all = ExerciseType.values().toList()
                val available = lastExerciseType?.let { last -> all.filter { it != last } } ?: all
                available.random()
            }
        }
    }

    private fun makeFillBlank(verse: MemorizedVerseEntity): ExerciseState.FillBlank {
        val words = verse.verseText.split(" ")
        val count = words.size
        val blankCount = when {
            count < 10 -> 1
            count < 20 -> 2
            else -> 3
        }
        val step = maxOf(1, count / (blankCount + 1))
        val blankIndices = mutableSetOf<Int>()
        for (i in 1..blankCount) {
            var idx = minOf(i * step, count - 1)
            for (offset in 0 until 5) {
                val candidate = (idx + offset) % count
                if (words[candidate].count { it.isLetter() } > 3) {
                    idx = candidate
                    break
                }
            }
            blankIndices.add(idx)
        }
        val segments = words.mapIndexed { i, word ->
            if (blankIndices.contains(i)) FillBlankSegment.Blank(word)
            else FillBlankSegment.Word(word)
        }
        return ExerciseState.FillBlank(verse, segments)
    }

    private fun makeWordDrag(verse: MemorizedVerseEntity): ExerciseState.WordDrag {
        val words = verse.verseText.split(" ")
        var shuffled = words.shuffled()
        if (shuffled == words && words.size > 1) shuffled = words.shuffled()
        return ExerciseState.WordDrag(verse, shuffled, words)
    }

    private fun makeMultipleChoice(
        verse: MemorizedVerseEntity,
        allVerses: List<MemorizedVerseEntity>
    ): ExerciseState.MultipleChoice {
        val distractors = allVerses
            .filter { it.id != verse.id }
            .shuffled()
            .take(3)
            .map { it.verseText }
        val fallbacks = listOf(
            "For the wages of sin is death; but the gift of God is eternal life through Jesus Christ our Lord.",
            "For all have sinned, and come short of the glory of God;",
            "That if thou shalt confess with thy mouth the Lord Jesus, and shalt believe in thine heart that God hath raised him from the dead, thou shalt be saved."
        )
        val options = mutableListOf(verse.verseText)
        options.addAll(distractors)
        var fi = 0
        while (options.size < 4) { options.add(fallbacks[fi++ % fallbacks.size]) }
        return ExerciseState.MultipleChoice(verse, options.take(4).shuffled())
    }
}
